Add redux aa1 to improve basics aa3

schettn opened the pull request:
**Checklist:**

 <!--- Make sure your PR is documented and tested before submission. Put an `x` in all the boxes that apply: -->
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change?
 - [x] Have you tested your changes with successful results?


**Type of Changes:**

 <!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] New feature (non-breaking change which adds functionality)
 - [ ] Documentation (non-breaking change which adds documentation)
 - [x] Breaking change (fix or feature that would cause existing functionality to change)


**What is the current behavior? (link to any open issues here)**
- Currently the sign in button in the navigation bar redirects to the start page, but not directly to the sing in form. #73
- Currently it is not possible to upload custom avatar images. #76 
- Currently user settings can be changed via a modal. #77 
- Currently when recaching data the avatar image resets. #78 
- Currently the application structure does not fit our needs because several issues. #80
- Currently there is no notification when sining in fails. #81 
- Currently the pinned repositories are kind of useless because they do not show any useful information. #82 
- Currently there are some issues in styling when displaying projects. #83 
- Currently there is no language chart for projects. #84 
- Currently there is no default image for projects without an avatar image/owner. #85 
- Currently there is no language chart for a user. #86

**What is the new behavior (if this is a feature change)?**
- Now it is possible to upload custom avatar images. #76 
- Now user settings can be changed via a separate settings page. #77 
- Now when uploading a avatar image it is permanent. #78 
- Now redux is implemented which resolves issues like active tabs when state changes,.. #80 
- Now when singing in fails there is a notification in the login form. #81 
- Now the pinned repositories are temporary disabled. #82 
- Now the styling of projects is refined and also works fine with projects without an owner. #83 
- Now there is a language chart for each project. #84 
- Now there is a default image for projects without a an avatar/owner. #85
- Now there is a language chart on each user profile. #86


**Other information**:
- 🐍

schettn committed:
Upgrade to redux

New redux structure was roughly implemented. Refinements take place in the following commits. But .. redux is fun :)

schettn committed:
Enhance Redux

Some bugs are fixed and the actions were adjusted.

schettn committed:
Improve user fetching

Incorrect user fetching has been fixed.

pinterid committed:
Update package.json

React-Avatar-Editor was added for editing images on the site.

pinterid committed:
Add profile picture modal

The modal for editing images was added.

pinterid committed:
Update modal export/imports

The modal imports and exports were updated.

pinterid committed:
Add settings page

A page for editing user settings was added.

pinterid committed:
Implement settings page

The path for the settings page was set.

schettn committed:
Upgrade registration component with redux

The registration works now with the new structure.

pinterid committed:
Update navbar settings

The navbar settings now work with the settings page.

pinterid committed:
Update settings action

The settings action is now capable of updating profile pictures.

schettn committed:
Login after regisration

A context error where the login function was not defined due to asynchronously
is fixed now. After registering a user will be logged in.

schettn committed:
Improve settings page accesability

Now it is possible to access the settings page directly by entering the url. Access is only granted to the own settings page.

schettn committed:
Improve user menu

Now when clicking on `My Profile` only the corresponding components are rendered and not longer everyone like before.

schettn committed:
Remove legacy code

Due to the switch to Redux, a lot of code is obsolete and is now being removed.

schettn committed:
Merge branch 'improve-settings-aa1' into add-redux-aa1

schettn committed:
Improve Redux DevTools integration

Now  Redux DevTools only applies in development mode.

schettn committed:
Fix syntax error of previous merge

The resolving of the merge conflicts went wrong and produced a syntax error which
is fixed now.

schettn committed:
Conditional Rendering of Talk Interaction Buttons

Uploading and deleting of talks is now only possible if a user is signed in
and on his profile.

pinterid committed:
Remove init blank

Blank initialization was removed because it is no longer needed.

pinterid committed:
Add input value check

The check was added to prevent the uncontrolled component warning.

schettn committed:
Merge remote-tracking branch 'origin/improve-settings-aa2' into add-redux-aa1

Aichnerc committed:
Add Language chart

The chart for displaying languages has been added.

Aichnerc committed:
Improve projects

The project component has been improved and now actually looks nice.

Aichnerc committed:
Improve profile info

The language chart and top languages have been implemented.

Aichnerc committed:
Disable pinned

The pinned posts have been temporarily disabled.

schettn committed:
Merge branch 'improve-design-ba1' into add-redux-aa1

pinterid committed:
Update autch actions

The auth actions errors were updated.

pinterid committed:
Update login form errors

The login form errors were updated.

schettn committed:
Merge branch 'add-notifications-aa1' into add-redux-aa1

schettn committed:
Add Default Project Avatar

When there is no image available for a project, a default image is set.

schettn committed:
Improve Default Major Language Name

The default text has been adjusted cause of the suck-level of the previous one.

schettn committed:
Add Todo To LanguageChart Component

A todo has been added. It includes renaming of component and descriptions
due to a copy paste issue...

schettn committed:
Merge branch 'improve-design-ba2' into add-redux-aa1

schettn commented:
@snek-at/frontend Just have a look on logically errors or some other invalid stuff. Issues on console logs and/or comments like sections are not part of this PR. 

**This is taken care of in implement PR!**

pinterid commented:
Opening the settings page without being logged in should redirect to the login page.

Currently there is just an infinity loading screen. 
![image](https://user-images.githubusercontent.com/55298934/86578007-6aafa500-bf7b-11ea-91a5-b1a75e7a6134.png)

pinterid commented on src/components/atoms/Project/index.jsx:
```diff
@@ -20,30 +22,33 @@ class Project extends React.Component {
+                    : "https://img.favpng.com/4/1/20/computer-icons-gregorio-lvarez-museum-png-favpng-xPCKYGtUJihZeAkdqind0HH2N.jpg"
+                }
+              />
+              <small>/{repo.owner.username}</small>
```
```suggestion
              <small>{repo.owner.username}</small>
```

pinterid reviewed (approved):
I did not found any logical issues in this code, beside the ones i already marked.

schettn committed:
Update SignIn Error Message

The message have been refined cause the old one was a bit shitty.

schettn committed:
Add Redirect to Homepage in Settings page for unlogged users

Now when a anonymous user visits the settings page he is redirected to
the homepage with active login form.

schettn commented:
I have updated the login error message cause it was not very neat to me.

kleberbaum commented on src/components/molecules/Navbar/index.jsx:
```diff
@@ -59,145 +59,137 @@ class Navbar extends React.Component {
+                    <MDBDropdown>
+                      <MDBDropdownToggle nav caret>
+                        <img
+                          src={loggedUser.avatarUrl}
```
This doesn't seem to work yet.
![image](https://user-images.githubusercontent.com/26285351/86581902-0db6ed80-bf81-11ea-957a-32c782bb454b.png)

pinterid replied:
This will work after the backend update, currently the logged in user only has a valid username. The profile picture is set static.

kleberbaum reviewed (approved):
Should be ok to merge, but you can also fix this issue beforehand.

pinterid committed:
Fix Codacy

Issues marked by Codacy were resolved.

pinterid committed:
Fix Codacy

Issues marked by Codacy were resolved.

pinterid committed:
Fix Codacy

Codacy errors were fixed.

schettn committed:
Move Changeable Content to User Section

Previously some data was stored in the profile section. Therefore when adjusting
data in settings it got overwritten during a refetch.
By moving this data to user section it does not get overwritten anymore.

schettn committed:
Add Optional Chaining to Login Result

Now when the result is undefined nothing happens cause there is no error.

schettn committed:
Fix Redirect to Homepage in Settings Page

There was an error where it always redirected if the settings url was entered
directly. This issue is fixed now,

schettn committed:
Remove Workaround

There was a workaround to update user settings. This is removed now because it
resets the user data every read cache iteration.
